docs/design/maven.md

```diff
@@ -0,0 +1,141 @@
+
+### Maven purls
+
+Purls for `pkg:maven` ([docs][maven-purl]) fully describe the location of an artifact in a maven repository. That is because
+all parts of the purl (`namespace`, `name`, `version` and the `classifier` and `type` qualifiers) are used to construct
+the download url for the artifact. [Maven documentation][maven-repo-layout] describes how to turn these coordinates into
+an actual download url. Here's an example:
+
+**Purl:**
+
+```
+pkg:maven/ga.io.quarkus/quarkus-core@3.8.5.redhat-00004?type=jar&repository_url=https://maven.repository.redhat.com&checksums=sha1:e4ca5fadf89e62fb29d0d008046489b2305295bf
+```
+
+**Mapping to GAV:**
+
+```
+groupId: ga.io.quarkus
+artifactId: quarkus-core
+version: 3.8.5.redhat-00004
+classifier: null
+type: jar
+```
+
+Using `classifier` and `type` we can determine the `extension` is `jar` ([docs][maven-extension]).
+
+**Download url**:
+
+Template:
+
+```
+${groupId as directory}/${artifactId}/${baseVersion}/${artifactId}-${version}.${extension}
+```
+
+Resulting url:
+
+```
+https://maven.repository.redhat.com/ga/io/quarkus/quarkus-core/3.8.5.redhat-00004/quarkus-core-3.8.5.redhat-00004.jar
+```
+
+## Design
+
+In this section, I will try to cover individual parts of the design.
+
+### Source repository (cachi2 lockfile)
+
+Currently, the cachi2 lockfile only allows `pkg:generic` purl type. This design will introduce a new purl type `pkg:maven`
+to support pre-fetching individual maven artifacts. The lockfile currently supports specifying the file to be fetched either
+as a purl or as a pair of `download_url` and `checksums`. Maven artifacts should only be specified as a purl in order to
+be reported as such in the output SBOM. Here's an example of a lockfile that fetches a maven artifact:
+
+```yaml
+metadata:
+  version: 1.0.0
+artifacts:
+  - purl: pkg:maven/ga.io.quarkus/quarkus-core@3.8.5.redhat-00004?type=jar&repository_url=https://maven.repository.redhat.com&checksums=sha1:e4ca5fadf89e62fb29d0d008046489b2305295bf
+    target: quarkus.jar
+```
+
+### Integration with existing cachi2 features
+
+As mentioned above, the lockfile format is identical to the one used for describing generic artifacts, therefore it makes
+sense to extend the existing lockfile with support for maven purls. The only difference would be in the output SBOM where
+maven artifacts would be reported as `pkg:maven`, instead of `pkg:generic`.
+
+There is an open question of how the UX should be. I will try to list the available options with some pros and cons.
+
+#### 1) Completely encapsulated in generic package manager
+
+This option would just extend the lockfile with support for maven purls, meaning the user would still be calling just the
+generic package manager. The benefit of this approach would be simplicity, as it would one package manager per lockfile.
+On the other hand, maven artifacts are not so generic.
+
+Example:
+```
+cachi2 fetch-deps --source /path/to/sources '[{"type": "generic"}]'
```

Comment:
Thank you for the example! There is no mention of a lock file here. Does this mean that a lock file will always reside in same location? Is there an agreement on where to put it and who would do that?

Reply:
At this time, yes. The generic lockfile is assumed to be present in the root of the cloned input repository. Here's a [link to the code](https://github.com/containerbuildsystem/cachi2/pull/681/files#diff-e65d82e684cdd28151a5ddccb300c06882c33725bc751263d1ffb3a0f1b878eaR14)

Reply:
Who will put it there?

Reply:
Not sure I understand where this question is going. It should be commited directly in the repository. An user that wants to use cachi2's maven artifact feature would do that